.github/workflows/test.yml

```diff
@@ -1,23 +1,24 @@
 name: test suite
 
-on:
-  push:
-    branches: [main]
-  pull_request:
+on: [push, pull_request, workflow_dispatch]
```

Comment:
This would result in the test suite being triggered twice when a pull request is made from the main repo. That's why I singled out the main branch.

Reply:
The status quo means I need to edit the yml in my fork, and then have to remember to revert it before creating a PR, which is an extra impediment to contributing.

Or I can submit a PR on my branch first to check it passes -> extra impediment.

Or I can develop on my `main` -> not a best practice, we should use feature branches, and I only have a single `main` -> extra impediment.

For upstream, I think the extra 2 minutes for the push build is acceptable.

CIs allow us to test on a much larger matrix of Python versions and operating systems, much more than is possible locally and sometimes faster too. I sometimes avoid contributing to projects with no/broken CI. 

I believe the use of CIs are of huge benefit for software quality and development, especially when so easy nowadays; Travis CI really revolutionised this, remember how hard testing was before. We should rather encourage people to test on their own forks _before_ creating PRs.

Fail fast: shorten the time to finding a failure, enable quicker loops of incremental development.

Alternatively, we can do something along the lines of:

```yml
# We want to run on external PRs, but not on our own internal PRs as they'll be run
# by the push to the branch.
if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
```

https://github.com/has2k1/plotnine/blob/e2ea6c9e731454860619fa79e28954ca9ea3e2f7/.github/workflows/testing.yml#L9-L11

Reply:
> The status quo means I need to edit the yml in my fork, and then have to remember to revert it before creating a PR, which is an extra impediment to contributing.

Why? My expectation is that you create a draft PR in case there is reasonable doubt that local testing would not reveal all the issues. No need to edit the yml. This has worked very well in my existing projects. The only downside is that maintenance branches need to be include on the branch list.

> I believe the use of CIs are of huge benefit for software quality and development, especially when so easy nowadays; Travis CI really revolutionised this, remember how hard testing was before. We should rather encourage people to test on their own forks before creating PRs.

There's no need to convince me of the benefits of CI. As an aside, Travis CI never had the problem with duplicate runs that GitHub does.

Reply:
To catch those unexpected failures. When there's not reasonable doubt. This can often be for older/newer Python versions (or, not relevant here, path things on Windows). Even to test changes made from a mobile. Testing feature branches on pushes has worked very well on my projects :)

Let's have the machines work for us. Anyway, reverted!